File: Gold_1/백준_2042_구간합구하기.cpp
#include<bits/stdc++.h>
using namespace std;
typedef long long ll;
ll N , M , K;
const ll MAX = 1e6;
ll num[MAX];
ll prefix_sum[4 * MAX];

ll Prefix_Sum_Tree_Init(ll st , ll en , ll n)
{
    if (st == en)
        return prefix_sum[n] = num[st];
    ll mid = (st + en) / 2;
    return prefix_sum[n] = Prefix_Sum_Tree_Init(st , mid , 2 * n) + Prefix_Sum_Tree_Init(mid + 1 , en , 2 * n + 1);
}

ll Prefix_Sum(ll st , ll en , ll n , ll l , ll r)
{
    if (st > r || en < l)
        return 0;
    if (st >= l && en <= r)
        return prefix_sum[n];
    ll mid = (st + en) / 2;
    return Prefix_Sum(st , mid , 2 * n , l , r) + Prefix_Sum(mid + 1 , en , 2 * n + 1 , l , r);
}

void Prefix_Update(ll st , ll en , ll n , ll idx , ll val)
{
    if (idx < st || en < idx)
        return;
    prefix_sum[n] += val;
    if (st == en)
        return;
    ll mid = (st + en) / 2;
    Prefix_Update(st , mid , 2 * n , idx , val);
    Prefix_Update(mid + 1 , en , 2 * n + 1 , idx , val);
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cin >> N >> M >> K;
    for (ll i = 0;i < N;i++)
        cin >> num[i];
    Prefix_Sum_Tree_Init(0 , N - 1 , 1);
    for (ll i = 0;i < M + K;i++)
    {
        ll a , b , c;
        cin >> a >> b >> c;
        if (a == 1)
        {
            ll tmp = c - num[b - 1];
            num[b - 1] = c;
            Prefix_Update(0 , N - 1 , 1 , b - 1 , tmp);
        }
        else if (a == 2)
            cout << Prefix_Sum(0 , N - 1 , 1 , b - 1 , c - 1) << "\n";
    }
}

File: Gold_1/백준_2042_구간합구하기.py
import sys
input = sys.stdin.readline
N, M, K = map(int, input().split())
num = [int(input().rstrip()) for i in range(N)]
sum_tree = [0]*(4*N)


def Sum_Tree_Init(st, en, n):
    if st == en:
        sum_tree[n] = num[st]
        return sum_tree[n]
    mid = (st + en) // 2
    sum_tree[n] = Sum_Tree_Init(st, mid, 2*n) + \
        Sum_Tree_Init(mid + 1, en, 2*n + 1)
    return sum_tree[n]


def Prefix_Sum(st, en, n, l, r):
    # st,en : 현재 보고있는 인덱스
    # n : 구간 합 트리의 인덱스
    # l,r : 구간 합을 구하고자 하는 범위
    # 만약 구하고자 하는 범위를 벗어나면 0 리턴
    if l > en or r < st:
        return 0
    # 구하고자 하는 범위가 완전히 속하면 구간 합 트리의 값 리턴
    if l <= st and en <= r:
        return sum_tree[n]
    # 벗어난 범위가 있다면 두 부분으로 나누어서 구간 합을 구한다.
    mid = (st + en) // 2
    return Prefix_Sum(st, mid, 2*n, l, r) + Prefix_Sum(mid + 1, en, 2*n + 1, l, r)


def Update_Value(st, en, n, idx, v):
    if idx < st or idx > en:
        return
    sum_tree[n] += v
    if st == en:
        return
    mid = (st + en) // 2
    Update_Value(st, mid, 2*n, idx, v)
    Update_Value(mid + 1, en, 2*n + 1, idx, v)


Sum_Tree_Init(0, N - 1, 1)
for i in range(M + K):
    a, b, c = map(int, input().split())
    if a == 1:
        # b번째 수를 c로 바꾼다.
        tmp = c - num[b - 1]
        num[b - 1] = c
        Update_Value(0, N - 1, 1, b - 1, tmp)
    elif a == 2:
        # b~c 번째 수까지 합을 구해서 출력한다.
        print(Prefix_Sum(0, N - 1, 1, b - 1, c - 1))
